takashi-market-maker.md

```diff
@@ -0,0 +1,170 @@
+# Market Maker
+
+## Algorithm
+
+> Write some code to create an algorithm that would maximize the amount of
+> money, assuming that nobody else buys and sells.
+
+some pseudocode:
+
+```py
+for order in orders:
+  curr_asset_price = lookup_asset_price(order[asset])
+  if order[buy] and order[price] >= curr_asset_price * 1.01:
+    execute_sell_order(order)
+  elif order[sell] and order[price] <= curr_asset_price * .99:
+    execute_buy_order(order)
+```
+
+> What happens if this becomes an n-sided market, where there are many
+> market makers? How does this change the problem?
+
+If this becomes a n-sided problem, speed matters quite a lot -- if you
+wait, someone else will snatch up the trade. Also, with more
+competition, the spread between bid and ask might have to tighten, but
+if there's less competition, the bid and ask spread can widen. You can
+code up an algorithm that adjusts automatically to market volume in
+this case.
+
+> Assume you have a limited amount of capital, which you don't want to
+> expend all at once (since there's volatility in asset prices, which may
+> go up +-10% a day), and there might be a larger amount of sell offers
+> than buy offers, for a particular asset. What happens if a stock goes up
+> too much, or down too much in a given day?
+
+If a stock goes up too much or too little in a given day, as a market
+maker, you'd like to avoid the stock (since volatility decreases money
+earned). You might try to calculate volatility and avoid volatile
+assets.
+
+## System Design
+
+> Design a system that listens to a server that sends requests, records
+> them to a database, and executes or passes on each offer, based on if it
+> is being sold for a 1% discount or if someone is willing to buy at a 1%
+> premium.
+
+First, let's ask about durability guarantees. Since we're a taxable
+financial entity, for tax purposes, we'll need to persist every
+executed order to stable storage, like a database.
+
+We can imagine that we would **receive** a potential offer, do some
+logic to check if we'd like to execute, and if so, send an **execute**
+back to the server. The server will then respond back to us with a
+response that tells us whether or not we were fast enough. If we were,
+we can record the result in the database. The database schema might
+look like this:
+
+```sql
+create table transactions (
+  id INTEGER PRIMARY KEY AUTO INCREMENT,
+  created_at DATETIME NOT NULL DEFAULT NOW(),
+  asset_id INTEGER FOREIGN KEY REFERENCES ASSETS.ID,
+  price MONEY NOT NULL,
+  action ENUM { BUY | SELL } NOT NULL,
```

Comment:
this is fine but how does it interface with your hashmap?

Reply:
This doesn't interface with the hashmaps. The hashmaps are in-memory for fast access, and choosing which offers to buy, but these transactions just record the offers we won, for analysis later by traders/analysts. 

We want to avoid reading from the database much since keeping everything in-memory speeds it all up.

Reply:
Makes sense